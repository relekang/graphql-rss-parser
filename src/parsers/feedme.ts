import _debug from 'debug';
import { Readable } from 'stream';
import FeedMe from 'feedme';
import { FeedObject } from 'feedme/dist/parser';

import { EmptyParserOutputError, ParserError } from '../errors';
import { Item, ParserResponse } from '../types';

const debug = _debug('graphql-rss-parser:parsers:feedme');

const findHtmlLink = (array: FeedObject[]): string | undefined => {
  const link = array.find(
    (item) =>
      typeof item === 'object' &&
      item['rel'] === 'alternate' &&
      item['type'] === 'text/html'
  );
  if (typeof link === 'object' && typeof link?.['href'] === 'string') {
    return link?.['href'] || undefined;
  }
  return undefined;
};

function evaluateLink(link: FeedObject | FeedObject[] | undefined): string {
  if (Array.isArray(link)) {
    const htmlLink = findHtmlLink(link);
    if (htmlLink) {
      return htmlLink;
    }
  }
  if (typeof link === 'string') {
    return link;
  }
  throw new ParserError(new Error('Missing link'), 'FEEDME');
}

function unpack(
  input: FeedObject | FeedObject[] | undefined,
  attribute: string,
  required: true,
  key: string
): string;
function unpack(
  input: FeedObject | FeedObject[] | undefined,
  attribute: string,
  required?: false
): string | undefined;
function unpack(
  input: FeedObject | FeedObject[] | undefined,
  attribute: string,
  required?: boolean,
  key?: string
): string | undefined {
  let output = undefined;
  if (input && Array.isArray(input)) {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    output = unpackArray(input, attribute)[0];
  }
  if (typeof input === 'string') {
    output = input;
  }
  // @ts-ignore ---
  if (typeof input === 'object' && typeof input[attribute] === 'string') {
    // @ts-ignore ---
    output = input[attribute] as string | undefined;
  }

  if (required && !output) {
    throw new ParserError(new Error(`Missing field ${key}`), 'FEEDME');
  }

  return output;
}

function unpackArray(
  input: FeedObject | FeedObject[] | undefined,
  attribute: string
): string[] {
  if (Array.isArray(input)) {
    return input
      .map((item) => unpack(item, attribute))
      .filter((item): item is string => !!item);
  }
  if (typeof input === 'string') {
    const unpacked = unpack(input, attribute);
    return unpacked ? [unpacked] : [];
  }
  return [];
}

export function parse(feed: string): Promise<ParserResponse> {
  return new Promise((resolve, reject) => {
    debug('starting to parse');
    try {
      if (feed.includes('medium.com')) {
        throw new Error('Failed to parse');
      }

      const parser = new FeedMe(true);

      parser.on('end', () => {
        const parsed = parser.done();
        if (!parsed) {
          return reject(new EmptyParserOutputError());
        }
        debug('done parsing');
        try {
          resolve({
            parser: 'FEEDME',
            title: unpack(parsed['title'], 'text', true, 'title'),
            description: unpack(parsed['description'], 'text'),
            home_page_url: evaluateLink(parsed['link']),
            feed_url: undefined,
            items: parsed.items.map((item): Item => {
              const pubDate = unpack(item['pubdate'], 'text');
              return {
                title: unpack(item['title'], 'text'),
                url: evaluateLink(item['link']),
                id: unpack(item['id'] || item['guid'], 'text', true, 'id'),
                content_html: unpack(item['description'], 'text'),
                tags: unpackArray(item['category'], 'text'),
                date_published: pubDate
                  ? new Date(pubDate).toISOString()
                  : pubDate,
                authors: unpack(item['author'], 'name')
                  ? [{ name: unpack(item['author'], 'name') }]
                  : [],
              };
            }),
          });
        } catch (error: any) {
          reject(new ParserError(error, 'FEEDME'));
        }
      });

      parser.on('error', (error) => {
        debug('parsing failed with error', error);
        reject(new ParserError(error, 'FEEDME'));
      });

      const stream = new Readable();
      stream.pipe(parser);
      stream.push(feed);
      stream.push(null);
    } catch (error) {
      debug('parsing failed with error', error);
      reject(error);
    }
  });
}
