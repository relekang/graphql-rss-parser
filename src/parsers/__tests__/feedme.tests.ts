/* eslint-env jest */
import { parse } from '../feedme'
import mockAxios from '../../__tests__/mockAxios'

test('should parse string from rolflekang.com/feed.xml', async () => {
  const fixture = await mockAxios({ url: 'https://rolflekang.com/feed.xml' })

  expect(await parse(fixture.data)).toMatchSnapshot()
})

test('should parse string from google.blogspot.com/feeds/posts/default', async () => {
  const fixture = await mockAxios({ url: 'http://google.blogspot.com/feeds/posts/default' })

  expect(await parse(fixture.data)).toMatchSnapshot()
})
