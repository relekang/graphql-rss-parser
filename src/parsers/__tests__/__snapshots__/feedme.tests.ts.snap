// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`should parse string from google.blogspot.com/feeds/posts/default 1`] = `
{
  "description": undefined,
  "feed_url": undefined,
  "home_page_url": "http://google.blogspot.com/",
  "items": [
    {
      "authors": [
        {
          "name": "Brett Wiltshire",
        },
      ],
      "content_html": undefined,
      "date_published": undefined,
      "id": "tag:blogger.com,1999:blog-2120328063286836889.post-7340716506563491347",
      "tags": [],
      "title": "Google to Acquire Blogger",
      "url": "http://google.blogspot.com/2011/04/google-to-acquire-blogger.html",
    },
  ],
  "parser": "FEEDME",
  "title": "Google on BlogSpot",
}
`;

exports[`should parse string from rolflekang.com/feed.xml 1`] = `
{
  "description": "undefined",
  "feed_url": undefined,
  "home_page_url": "https://rolflekang.com",
  "items": [
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2020-11-08T00:00:00.000Z",
      "id": "https://rolflekang.com/using-certbot-with-ansible",
      "tags": [
        "letsencrypt",
        "acme",
        "certbot",
        "ansible",
      ],
      "title": "Using certbot with Ansible",
      "url": "https://rolflekang.com/using-certbot-with-ansible",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2020-09-12T00:00:00.000Z",
      "id": "https://rolflekang.com/ansible-handlers-in-loops",
      "tags": [
        "ansible",
        "devops",
        "ops",
        "infrastructure-as-code",
      ],
      "title": "Using Ansible handlers in loops",
      "url": "https://rolflekang.com/ansible-handlers-in-loops",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2020-05-26T00:00:00.000Z",
      "id": "https://rolflekang.com/serving-plain-text-with-nextjs",
      "tags": [
        "nextjs",
        "javascript",
        "curl",
        "nodejs",
      ],
      "title": "Serving text/plain for curl with Next",
      "url": "https://rolflekang.com/serving-plain-text-with-nextjs",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2020-05-03T00:00:00.000Z",
      "id": "https://rolflekang.com/wireless-uplinks-with-unifi",
      "tags": [
        "networking",
        "unifi",
        "ubiquiti",
        "homelab",
      ],
      "title": "Wireless uplinks with Unifi",
      "url": "https://rolflekang.com/wireless-uplinks-with-unifi",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2019-11-14T00:00:00.000Z",
      "id": "https://rolflekang.com/using-git-commits-instead-of-stash",
      "tags": [
        "git",
      ],
      "title": "Using git commits instead of git stash",
      "url": "https://rolflekang.com/using-git-commits-instead-of-stash",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2019-02-26T00:00:00.000Z",
      "id": "https://rolflekang.com/twitter-cards-for-gatsby-posts",
      "tags": [
        "gatsby",
        "canvas",
        "javascript",
      ],
      "title": "Twitter cards for Gatsby posts",
      "url": "https://rolflekang.com/twitter-cards-for-gatsby-posts",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2019-02-20T00:00:00.000Z",
      "id": "https://rolflekang.com/related-articles-with-gatsby",
      "tags": [
        "gatsby",
        "javascript",
      ],
      "title": "Related articles with Gatsby",
      "url": "https://rolflekang.com/related-articles-with-gatsby",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2019-02-12T00:00:00.000Z",
      "id": "https://rolflekang.com/creating-a-cli-with-reason-native",
      "tags": [
        "reasonml",
        "native",
      ],
      "title": "Creating a CLI with Reason native",
      "url": "https://rolflekang.com/creating-a-cli-with-reason-native",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2017-05-28T00:00:00.000Z",
      "id": "https://rolflekang.com/testing-simple-graphql-services",
      "tags": [
        "graphql",
        "testing",
        "javascript",
      ],
      "title": "Testing simple GraphQL services",
      "url": "https://rolflekang.com/testing-simple-graphql-services",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2017-01-02T00:00:00.000Z",
      "id": "https://rolflekang.com/lets-run-some-races",
      "tags": [
        "running",
      ],
      "title": "Let's run some races this year",
      "url": "https://rolflekang.com/lets-run-some-races",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2017-01-01T00:00:00.000Z",
      "id": "https://rolflekang.com/software-development-on-an-ipad",
      "tags": [
        "ipad",
        "development-environment",
      ],
      "title": "Software development on an iPad",
      "url": "https://rolflekang.com/software-development-on-an-ipad",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2016-06-10T00:00:00.000Z",
      "id": "https://rolflekang.com/filtering-lint-errors",
      "tags": [
        "linter",
        "linting",
        "development",
      ],
      "title": "Filtering lint errors",
      "url": "https://rolflekang.com/filtering-lint-errors",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2015-07-26T00:00:00.000Z",
      "id": "https://rolflekang.com/interrail-summer",
      "tags": [
        "travel",
        "interrail",
      ],
      "title": "Interrail summer",
      "url": "https://rolflekang.com/interrail-summer",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2015-05-23T00:00:00.000Z",
      "id": "https://rolflekang.com/writing-latex-in-atom",
      "tags": [
        "atom",
        "writing",
        "procrastination",
      ],
      "title": "Writing (latex) in Atom",
      "url": "https://rolflekang.com/writing-latex-in-atom",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2014-12-30T00:00:00.000Z",
      "id": "https://rolflekang.com/rmoq",
      "tags": [
        "python",
        "testing",
        "mocking",
      ],
      "title": "rmoq - a request mock cache",
      "url": "https://rolflekang.com/rmoq",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2014-10-03T00:00:00.000Z",
      "id": "https://rolflekang.com/django-nopassword-one-point-o",
      "tags": [
        "django",
        "nopassword",
        "django-nopassword",
        "authentication",
      ],
      "title": "django-nopassword reached 1.0",
      "url": "https://rolflekang.com/django-nopassword-one-point-o",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2014-05-07T00:00:00.000Z",
      "id": "https://rolflekang.com/moving-to-pelican",
      "tags": [
        "this",
        "static-site",
        "pelican",
        "jekyll",
      ],
      "title": "Moving the blog to pelican",
      "url": "https://rolflekang.com/moving-to-pelican",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2014-02-02T00:00:00.000Z",
      "id": "https://rolflekang.com/building-tumblr-theme-with-grunt",
      "tags": [
        "gruntjs",
        "cookiecutter",
        "tumblr",
      ],
      "title": "Building a Tumblr theme with GruntJS",
      "url": "https://rolflekang.com/building-tumblr-theme-with-grunt",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2013-10-24T00:00:00.000Z",
      "id": "https://rolflekang.com/backup-routines-for-redisdb",
      "tags": [
        "db",
        "backup",
        "redis",
      ],
      "title": "Backup routine for Redis",
      "url": "https://rolflekang.com/backup-routines-for-redisdb",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2013-05-22T00:00:00.000Z",
      "id": "https://rolflekang.com/django-app-custom-user",
      "tags": [
        "django",
        "custom user",
      ],
      "title": "Making your Django app ready for custom users",
      "url": "https://rolflekang.com/django-app-custom-user",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2013-03-03T00:00:00.000Z",
      "id": "https://rolflekang.com/stats-are-fun",
      "tags": [
        "pypstats",
        "pypi",
        "stats",
        "geek-tool",
      ],
      "title": "Stats are fun",
      "url": "https://rolflekang.com/stats-are-fun",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2013-02-26T00:00:00.000Z",
      "id": "https://rolflekang.com/alfred-scripts-for-jekyll",
      "tags": [
        "alfredapp",
        "jekyll",
        "shell script",
      ],
      "title": "Alfred scripts for Jekyll",
      "url": "https://rolflekang.com/alfred-scripts-for-jekyll",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2013-01-27T00:00:00.000Z",
      "id": "https://rolflekang.com/use-array-in-postgresql",
      "tags": [
        "tags",
        "postgres",
        "array",
        "pgarray",
        "django",
      ],
      "title": "Use array in postgresql with django",
      "url": "https://rolflekang.com/use-array-in-postgresql",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2013-01-03T00:00:00.000Z",
      "id": "https://rolflekang.com/apply-puppet-configuration-automatically",
      "tags": [
        "puppet",
        "web.py",
        "nginx",
        "uwsgi",
        "github",
        "github-hooks",
      ],
      "title": "Apply puppet automatically",
      "url": "https://rolflekang.com/apply-puppet-configuration-automatically",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2012-12-02T00:00:00.000Z",
      "id": "https://rolflekang.com/readable-wikipedia",
      "tags": [
        "wikipedia",
        "css",
        "readability",
        "stylebot",
      ],
      "title": "Readable Wikipedia",
      "url": "https://rolflekang.com/readable-wikipedia",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2012-09-28T00:00:00.000Z",
      "id": "https://rolflekang.com/django-nopassword",
      "tags": [
        "django",
        "no-password",
        "email-authentication",
        "authentication",
      ],
      "title": "django-nopassword",
      "url": "https://rolflekang.com/django-nopassword",
    },
    {
      "authors": [],
      "content_html": undefined,
      "date_published": "2012-09-16T00:00:00.000Z",
      "id": "https://rolflekang.com/the-github-lamp",
      "tags": [
        "arduino",
        "github",
        "git",
        "web.py",
      ],
      "title": "The Github lamp",
      "url": "https://rolflekang.com/the-github-lamp",
    },
  ],
  "parser": "FEEDME",
  "title": "Writing by Rolf Erik Lekang",
}
`;
